### ISSUE TITLE
nodejs default version / outdated nodejs

### ISSUE THREAD

**Issue by gilligan** [NixOS/nixpkgs#13103]
Currently the default nodejs version is `nodejs-4.2.3` which is pretty ancient. There is `nodejs-5_x` which is currently at `nodejs-5.4.1` which is also not recent anymore. 

I wonder if there is any reason not to default nodejs to version 5.x and just keep `nodejs-4_x` and `nodejs-0_10` alongside ?

What would the requirements/implications for this be ? I suppose all `nodePackages.xxx` packages would have to be updated ? Is that a problem ? Are there any other reasons not to do this ?

**Comment by zimbatm** [comment]
Testing and manpower most certainly. We need more help !

**Comment by gilligan** [comment]
@zimbatm I am interested in having a more up to date/usable nodejs integration in nix but I lack the experience to drive it myself. I am however very willing to invest time.

I started to look into https://github.com/adnelson/nixfromnpm which in my opinion should replace npm2nix because it does a way better job.

So is someone with enough experience available to step in and advise what possible steps about updating nodejs packages would be ? You @zimbatm ? :)

**Comment by svanderburg** [comment]
I believe Node.js 4.3.x is the LTS version whereas 5.x are feature
releases. I still use 4.x for most of my projects.

I don't know what policy is best but making the LTS release the default
wouldn't sound so bad to me.

Moreover, I have no clue in what direction the 5.x branch would go. Maybe
they will do another major npm change that will break things.
Op 19 feb. 2016 08:51 schreef "Tobias Pflug" <notifications@github.com>:

**Comment by svanderburg** [comment]
By the way, I didn't know there was another tool effort that can be used to
generate Nix expressions from NPM packages. If somebody really likes doing
(and maintaining!) it in such a way that it can cope with all the
(potential) future changes, that would be very nice! From my own
perspective, getting a generator to work is fun, but coping with all the
breaking changes in NPM (such as peer dependencies, flat-dependency
structure, because they do not seem to know how things should work either)
is a real pain and makes me quite sad!

At the same time, I have recently done some work on npm2nix as well (my
reegnineering branch, not the "vanilla" npm2nix) because my employer gave
me some time for it and they want some of their stuff to be fixed as well.
I started a new branch (reengineering2) from scratch that takes a
fundamentally different approach (
https://github.com/svanderburg/npm2nix/tree/reengineering2).

The old reengineering branch worked fine for all my Node.js 4.x-based
projects, but its biggest disadvantage was that it relied on building
configuration files in the instantiation phase. In the reengineering2
branch, I got rid of this approach and I compose the entire dependency
graph in one derivation. Also, with npm 3.x's flat dependency structure the
old approach no longer makes sense.

I also managed to almost simulate npm 3.x (that comes with Node.js 5.x)
flat-dependency behaviour, but there are still a few subtle issues which I
am still investigating at the moment. The README.md file of reengineering2
version is also still missing at this point, which also makes it a bit
impractical to use.

Anyway,

To me it does not really matter what approach we use. The problem is that
the set of NPM packages in Nixpkgs is broken for a quite a while and we
want to fix it.

Besides fixing it, I think many packages that we currently include should
not be in Nixpkgs. Libraries in NPM projects are stateful -- depending on
the state of the includer, NPM decides to which version a dependency
resolves. So it makes no sense to include library dependencies of a NPM
project from Nixpkgs. Instead, people should use a generator for their
private projects to see to which versions dependencies resolve.

The only NPM packages that should be in Nixpkgs IMO are:

- Command-line utilities/end-user programs
- Library packages that are dependencies of non-NPM packages

**Comment by zimbatm** [comment]
I agree, it doesn't make sense to include the libraries unless the language has a tested LTS release of their package. Haskell has stackackage which provides a single plane of packages that all work together. But node, ruby, python, even worse go will never have a single set of all their repository's packages that work together. Fighting to find that common set ourselves, most of the time without the help of the language's tools is a waste of time.

What we want in nix is specific overrides, for example that the v8 npm package depends on libv8. That way the user generating his npm2nix doesn't have to worry about system dependencies most of the time.

**Comment by zimbatm** [comment]
@gilligan sure, hit me on IRC if you want some help. I might not respond quickly so make just to stay logged. The first thing we can do is to update the 5.x node release without changing the default.

**Comment by domenkozar** [comment]
This is now fixed in master (soon in 16.03)

**Comment by gilligan** [comment]
I hope you don't mind me reactivating this thread but I think it's better to fit this in here instead of creating a new ticket: At the end of April nodejs 6.0 will be released which will be the the next LTS release:

![lts-schedule](https://github.com/nodejs/LTS/raw/master/schedule.png)

Starting October version 6.0 will be the most recent active LTS release. My suggestion would be:

* Introduce 6.0 as `nodejs-6_x` once it is available
* Switch `nodejs` derivation to 6.0 in October

How does that sound ? /cc @lo1tuma
